### PR TITLE
Clone and JSON for geometries

### PR DESCRIPTION
**Pull request by dubejf**
**Add missing clone()**
Provide a clone method to the specialized geometries in src/extras/geometry, otherwise:
- the `.parameters` are not copied
- the `.type` of the clone is "Geometry" or "BufferGeometry"
- the cloned object is not an instance of the initial type. It loses its specialization and becomes a plain `Geometry` or `BufferGeometry`.

**ObjectLoader improvement**

Add more geometries to ObjectLoader.js and pass down missing parameters to constructors for `CircleGeometry` and `CylinderGeometry`.

**Miscellaneous**
Class deriving from `PolyhedronGeometry` defined  `this.parameters` all over the place. Standardize to defining the `.type` and `.parameters` after the `PolyhedronGeometry` constructor is called.

Also fix prototype inheritance chain for classes deriving from `PolyhedronGeometry`.

Add unit tests for it all.

### CODE REVIEW

File: src/extras/geometries/BoxGeometry.js [hunk 1]
```diff
@@ -124,4 +124,21 @@ THREE.BoxGeometry = function ( width, height, depth, widthSegments, heightSegmen
...
+	var geometry = new THREE.BoxGeometry(
+		this.parameters.width,
+		this.parameters.height,
+		this.parameters.depth,
+		this.parameters.widthSegments,
+		this.parameters.heightSegments,
+		this.parameters.depthSegments
+	);
+
+	THREE.Geometry.prototype.clone.call( this, geometry );
```
Comment (mrdoob) [inline src/extras/geometries/BoxGeometry.js]: I don't think this is needed...?

Comment (dubejf) [inline src/extras/geometries/BoxGeometry.js]: It depends. Are modifications to the non-parametrized properties of a `BoxGeometry` allowed without flattening?

This fix is backward compatible with respect to code like this:

```
var pxGeometry = new THREE.PlaneGeometry( 100, 100 );
pxGeometry.faces[ 0 ].vertexColors = [ light, shadow, light ];
... 
copy = pxGeometry.clone(); // "exact" copy
```

Any tweaks to `.vertices`, `.faces` or `.faceVertexUvs` are copied. There are other properties like `.name` and `.colors` that seem relevant but are not currently handled by `Geometry.clone()`.

Should this use case be supported?

**Comment by WestLangley** [comment]
What is the rule by which `clone()` takes an argument? Is it only in classes that have been extended?

**Comment by dubejf** [comment]
There is nothing preventing any `clone()` to take an optional destination as its first argument, but it has to be taken into account by the implementation. It removes the guarantee of starting from an empty destination object - the object must be cleared before new data is appended to it.

This version of `Geometry.prototype.clone( geometry )` assumes that the destination geometry is a newly constructed geometry. To be fully compliant `Geometry.prototype.clone()` should clear the `morphTargets`, etc.

So, `clone` with an argument should only be called from an extended class.

**Comment by dubejf** [comment]
If BoxGeometry should only be allowed to modify its parameters, maybe this would be a better interface to seal off the details of the geometry and enforce these constraints.

```
// Does not extend THREE.Geometry or have vertices of its own, but exposes a compatible interface.
THREE.BoxGeometry2 = function ( width, height, depth, widthSegments, heightSegments, depthSegments ) {

	// Private geometry sidekick.
	var geometry = new THREE.Geometry(); // Or as buffer geometry

	this.type = 'BoxGeometry';

	this.parameters = {
		width: width,
		height: height,
		depth: depth,
		widthSegments: widthSegments,
		heightSegments: heightSegments,
		depthSegments: depthSegments
	};

	this.applyParameters = applyParameters();
	this.toGeometry() = /// flatten, returns a parameterless THREE.Geometry copy
	this.toBufferGeometry() = /// flatten, returns a parameterless THREE.BufferGeometry copy

	function applyParameters ( ) {

		// create / re-generate geometry from parameters

	}

	Object.defineProperties( this, {
		name: {
			get: function () {
				return geometry.name;
			}
		},
		boundingBox: {
			get: function () {
				if ( geometry.boundingBox === null ) {

					geometry.computeBoundingBox();

				}
				return geometry.boundingBox;
			}
		},
		vertices: {
			get: function () {
				return geometry.vertices;
			},
		}
		// ...
	} );

};
```

Use:
```
var box = new THREE.BoxGeometry2( 1, 1, 1 );
new THREE.Mesh( box, material );
render();

...
box.parameters.width = 2;
box.applyParameters();
render();
```

**Comment by WestLangley** [comment]
> So, clone with an argument should only be called from an extended class.

To make sure I understand this correctly, the answer to my question is "yes"?

And cloning with an argument is used primarily by the library -- not the user -- and only when cloning objects which are instances of an extended class. Correct?

I have not seen any use cases at the application layer of passing an argument into `clone()`, with the exception of `THREE.UniformsUtils.clone()`.

**Comment by dubejf** [comment]
@WestLangley Yes, you understood correctly.

I based it off what I understood to be the current practice. If that's too confusing, there could be two methods:
- [Public API] `clone()` which takes no arguments and create a new object.

- [Internal use only] For classes that are extended, `baseClone()` that takes an argument - a newly created object of the extended kind. It would not create a new object, just copy the base properties into to provided object.

`THREE.Object3D`, `THREE.Light`, etc. would adopt this new pattern.

**Comment by WestLangley** [comment]
@dubejf Yes. For the public API, `clone()` should not have an argument, IMO.

The private method, ( `_clone( object )` ? ), can be implemented any way you think is best.

**Comment by dubejf** [comment]
@WestLangley  Is that what you had in mind?

**Comment by WestLangley** [comment]
@dubejf Sweet. I think this is great! Many thanks.

Do you think we can get away with omitting a warning to users who pass an argument into `clone()`?

**Comment by dubejf** [comment]
Sure, we can emit a warning.

What should we do about `Mesh` and `Object3D` ? They take a `recursive` flag. Previous prototype was `clone: function ( object, recursive )`, now it is `clone: function ( recursive )`. Check that `arguments.length < 2` ? Check that the first arg is `true`, `false` or `undefined`?

As long as we are standardizing the interfaces, should all `clone`'s accept `recursive` as their first and only argument (at least those extending `Object3D`) ? Currently, recursive === true is implied for all objects in the `Object3D` family, except `LOD`.

**Comment by WestLangley** [comment]
Gee, I never noticed the recursive flag before. I see it works well when cloning a hierarchical model. How handy!

I'm not yet sure of the answers to your questions, as I have been sidetracked by the following issue:

Because `recursive` is true by default, I am not sure `Mesh` needs the recursive arg -- or, if it does, then `Line`, for example, should have the recursive arg, too.... Note that currently, a hierarchy of lines does clone recursively.

File: src/extras/geometries/BoxGeometry.js (hunk 1 above)
Comment (mrdoob) [inline src/extras/geometries/BoxGeometry.js]: > Should this use case be supported?

I think it shouldn't. I think we should, instead, encourage "converting" to "Geometry" when modifying vertices. This should keep things simpler.

Comment (WestLangley) [inline src/extras/geometries/BoxGeometry.js]: @mrdoob What happens when you re-comment on a previously-outdated diff -- as I am now also doing? It is collapsed and not visible... I wonder if the OP is notified somehow?...

Comment (dubejf) [inline src/extras/geometries/BoxGeometry.js]: I am notified but the messages are collapsed in the thread.

**Comment by dubejf** [comment]
So, should we expose the recursive option "everywhere"?

**Comment by WestLangley** [comment]
> So, should we expose the recursive option "everywhere"?

I was beginning to come to the conclusion that the `recursive` option is only needed in `Object3D` -- but I am not certain...

**Comment by dubejf** [comment]
Ok, so no `recursive` except for `Object3D.clone()`. I removed the `recursive` from the `Mesh` interface.

I'm hesitant to add a warning when `clone()` is called with an argument. That's a lot of clutter and a lot of bytes added to the download for a uncommon use case. The changelog is probably a better place to disclose this change.

If you agree, the patch is done!

**Comment by WestLangley** [comment]
> I was beginning to come to the conclusion that the recursive option is only needed in Object3D

If you believe my conjecture is correct, then this looks good.

File: src/extras/geometries/BoxGeometry.js [hunk 2]
```diff
@@ -124,4 +124,21 @@ THREE.BoxGeometry = function ( width, height, depth, widthSegments, heightSegmen
...
+	var geometry = new THREE.BoxGeometry(
+		this.parameters.width,
+		this.parameters.height,
+		this.parameters.depth,
+		this.parameters.widthSegments,
+		this.parameters.heightSegments,
+		this.parameters.depthSegments
+	);
+
+	THREE.Geometry.prototype.cloneProperties.call( this, geometry );
```
Comment (mrdoob) [inline src/extras/geometries/BoxGeometry.js]: I still think we don't need these...

**Comment by WestLangley** [comment]
Did you miss some?

    THREE.CubeTexture.clone = function ( texture ) {

**Comment by dubejf** [comment]
@WestLangley @mrdoob Done.

**Comment by mrdoob** [comment]
Howe about `copy` instead of `cloneProperties`?

**Comment by WestLangley** [comment]
@mrdoob You mean `copyProperties`? That does seem more descriptive.

**Comment by mrdoob** [comment]
> @mrdoob You mean `copyProperties`? That does seem more descriptive.

Well, wouldn't you think that `copy` is copying the properties? What else would it copy?

**Comment by WestLangley** [comment]
Thinking.... `copy` seems to be reserved for `math` objects -- with the exception of `BufferGeometry` that has a `copy` method....

**Comment by dubejf** [comment]
A uglier name was picked because this was intended to be an internal method, to be used only as a helper for `clone` and to be called with freshly constructed objects only.

`copy` could work if you are okay with people calling this method directly from their code, which would probably be okay most of the time.

Some implementation of `copy` might need to be changed to accept dirty objects. For example, clearing arrays before pushing new data into them, as the patch already does in `Geometry`:

https://github.com/mrdoob/three.js/pull/6825/files#diff-a7251e9e389d43e7878065274e4d067dR1058

Beside this caveat, I'm fine with any name discussed so far.

**Comment by dubejf** [comment]
Rebased, and since @gero3 has added a `devDependencies` section in `package.json`, remove the copy of `lodash.js` from `test/unit/` and use the npm module instead.

As long a you do `npm install` once in the root directory to install the dependencies, the unit test will work as before.

**Comment by dubejf** [comment]
One more thing regarding "`cloneProperties`"...

It takes a destination object in argument, and the source of the copy is the `this` object: `src.cloneProperties( dest )`. This is the same convention used by `clone` before.

However, this is verse of `copy` in the Math object: `dest.copy( src )`.

Beside a new name, I assume that you'll what me to reverse the source and the destination?

**Comment by WestLangley** [comment]
> Beside a new name, I assume that you'll what me to reverse the source and the destination?

Yes, `object._copyFrom( source )` seems more natural to me.

Regarding the method name, personally I prefer using the underscore notation because it is intended to be a private method.

**Comment by dubejf** [comment]
Modified per @WestLangley suggestion.

**Comment by WestLangley** [comment]
Many thanks. Using the word "source" in the method signature is very helpful, IMO.

I expect @mrdoob will not like the method name. What is important to me is answering this question: Do we want users calling it? If generally not, then it needs to be clearly private. If there is a compelling use case at the application level, then it should be public.

**Comment by mrdoob** [comment]
> I expect @mrdoob will not like the method name. 

Yeah... I prefer just `.copy()` and I think it's fine if people use it.

**Comment by dubejf** [comment]
Renamed `._copyFrom()` to `.copy()`. There was already an equivalent `copy()` method in `BufferGeometry`, which I've consolidated.

>  I think it's fine if people use it.

Some implementations of `copy()` might need to be improved a bit. For instance, `THREE.LOD.prototype.copy()` should clear existing levels before pushing the cloned levels. Not a showstopper either way.

File: src/core/BufferGeometry.js [hunk 3]
```diff
@@ -1108,23 +1082,55 @@ THREE.BufferGeometry.prototype = {
...
+			this.addDrawCall( offset.start, offset.count, offset.index );
 
-			geometry.addDrawCall( offset.start, offset.count, offset.index );
+		}
+
+		return this;
+
+	},
+
+	copy: function ( source ) {
```
Comment (mrdoob) [inline src/core/BufferGeometry.js]: There seem to be two `copy` in `BufferGeometry` now?

File: examples/js/renderers/CanvasRenderer.js [hunk 4]
```diff
@@ -22,7 +22,7 @@ THREE.SpriteCanvasMaterial.prototype.clone = function () {
 
 	var material = new THREE.SpriteCanvasMaterial();
 
-	THREE.Material.prototype.clone.call( this, material );
+	THREE.Material.prototype.copy.call( material, this );
```
Comment (mrdoob) [inline examples/js/renderers/CanvasRenderer.js]: Shouldn't this be `material.copy( this )`?

File: src/core/BufferGeometry.js (hunk 3 above)
Comment (dubejf) [inline src/core/BufferGeometry.js]: Oops

File: examples/js/renderers/CanvasRenderer.js (hunk 4 above)
Comment (dubejf) [inline examples/js/renderers/CanvasRenderer.js]: Same thing but yes why not.

File: src/materials/LineBasicMaterial.js [hunk 5]
```diff
@@ -47,7 +47,7 @@ THREE.LineBasicMaterial.prototype.clone = function () {
 
 	var material = new THREE.LineBasicMaterial();
 
-	THREE.Material.prototype.clone.call( this, material );
+	THREE.Material.prototype.copy.call( material, this );
```
Comment (mrdoob) [inline src/materials/LineBasicMaterial.js]: There seems to be a bunch of these. Everything else looks great though!

**Comment by dubejf** [comment]
Done.

The branch conflict is with `ProxyGeometry` (gone in `dev`). Should be easy to resolve!

**Comment by mrdoob** [comment]
Resolved! Many thanks!

**Pull request merged**